File: android/app/src/test/kotlin/com/matrimpathak/ledger/LocalSmsParserTest.kt
package com.matrimpathak.ledger

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * JVM unit test (no Android framework dependency) for LocalSmsParser,
 * asserted against the exact same test/fixtures/sms_samples.json fixture
 * file the Dart test (test/services/sms/local_sms_parser_test.dart) uses —
 * this is what catches the two interpreters drifting apart, since both are
 * held to identical expected extraction for identical input.
 *
 * Paths are relative to the Gradle module's working directory
 * (android/app/), hence the "../../" to reach the repo root.
 */
class LocalSmsParserTest {

    private val rules = JSONObject(
        File("../../assets/sms_patterns/bank_patterns.json").readText()
    )
    private val parser = LocalSmsParser(rules)
    private val samples = JSONObject(
        File("../../test/fixtures/sms_samples.json").readText()
    ).getJSONArray("samples")

    @Test
    fun `every fixture sample extracts the expected fields and confidence tier`() {
        for (i in 0 until samples.length()) {
            val sample = samples.getJSONObject(i)
            val id = sample.getString("id")
            val body = sample.getString("body")
            val expected = sample.getJSONObject("expected")

            val result = parser.parse(body)

            assertEquals("[$id] amount", expected.optNullableDouble("amount"), result.amount)
            assertEquals("[$id] direction", expected.optNullableString("direction"), result.direction)
            assertEquals("[$id] paymentMethod", expected.optNullableString("paymentMethod"), result.paymentMethod)
            assertEquals("[$id] txnCategoryHint", expected.optNullableString("txnCategoryHint"), result.txnCategoryHint)
            assertEquals("[$id] referenceNumber", expected.optNullableString("referenceNumber"), result.referenceNumber)
            assertEquals("[$id] merchantCandidate", expected.optNullableString("merchantCandidate"), result.merchantCandidate)
            assertEquals("[$id] accountLastDigits", expected.optNullableString("accountLastDigits"), result.accountLastDigits)
            assertEquals("[$id] matchedRuleId", expected.getString("matchedRuleId"), result.matchedRuleId)

            when (expected.getString("tier")) {
                "high" -> assertTrue("[$id] expected high confidence, got ${result.confidence}", result.isHighConfidence)
                "medium" -> assertTrue("[$id] expected medium confidence, got ${result.confidence}", result.isMediumConfidence)
                "low" -> assertTrue("[$id] expected low confidence, got ${result.confidence}", result.isLowConfidence)
            }
        }
    }

    @Test
    fun `matching a known account last-digits adds the instrument bonus`() {
        val body = "INR 12000.00 debited from your account for NEFT transfer to A/C XX4321. Avl Bal INR 45000.00"
        val without = parser.parse(body)
        val withAccount = parser.parse(
            body,
            accounts = listOf(mapOf("id" to "acc-1", "lastSixDigits" to "564321")),
        )

        assertEquals("acc-1", withAccount.matchedAccountId)
        assertEquals(0.20, withAccount.confidence - without.confidence, 0.001)
    }

    @Test
    fun `matching a known payment mode last-digits adds the instrument bonus`() {
        val body = "Rs.4500.00 spent using your Credit Card XX9988 at a merchant on 12-08-26."
        val without = parser.parse(body)
        val withMode = parser.parse(
            body,
            paymentModes = listOf(mapOf("id" to "mode-1", "lastFourDigits" to "9988")),
        )

        assertEquals("mode-1", withMode.matchedPaymentModeId)
        assertEquals(0.20, withMode.confidence - without.confidence, 0.001)
    }

    // A shape no static rule in bank_patterns.json covers at all — it
    // deliberately avoids every rule-triggering keyword ("debited",
    // "credited", "upi", "credit card", "atm", "refund"), same as the
    // real HDFC "Sent"/"Received" format that motivated this tier.
    private val iciciTemplate = SmsTemplate(
        id = "tpl-icici-debit",
        bank = "ICICI Bank",
        bankCode = "ICICIB",
        transactionType = "bank_debit",
        direction = "debit",
        paymentMethod = "bankTransfer",
        txnCategoryHint = null,
        skeleton = "ICICI Alert: INR {amount} moved from A/C XX{acct_last4} to {merchant} on {date}. Ref {refno}.",
        templateConfidence = 0.92,
    )
    private val iciciBody = "ICICI Alert: INR 1250.50 moved from A/C XX7788 to BLINKIT on 19-Aug-26. Ref 445566778899."

    @Test
    fun `falls back to a matching template when no static rule fits`() {
        val result = parser.parse(iciciBody, sender = "AD-ICICIB-S", templates = listOf(iciciTemplate))

        assertEquals("template", result.matchedRuleId)
        assertEquals("tpl-icici-debit", result.matchedTemplateId)
        assertEquals(1250.5, result.amount!!, 0.001)
        assertEquals("debit", result.direction)
        assertEquals("bankTransfer", result.paymentMethod)
        assertEquals("445566778899", result.referenceNumber)
        assertEquals("BLINKIT", result.merchantCandidate)
        assertEquals("7788", result.accountLastDigits)
        assertTrue(result.isHighConfidence)
    }

    @Test
    fun `a template from a different bank's sender is not tried`() {
        val result = parser.parse(iciciBody, sender = "VM-HDFCBK-T", templates = listOf(iciciTemplate))

        assertEquals("none", result.matchedRuleId)
        assertNull(result.matchedTemplateId)
    }

    @Test
    fun `a static rule match takes priority over a template`() {
        // upi_debit already covers this shape via bank_patterns.json's own
        // rules — the template tier must never be consulted once a static
        // rule fires, so a bad/duplicate learned template can't override
        // hand-verified extraction.
        val body = "Rs.286.00 debited from A/C XX1234 to VPA rajesh@okhdfc on " +
            "15-08-26. UPI Ref No 402312345678. Not you? Call 1800123456"
        val decoy = SmsTemplate(
            id = "decoy",
            bank = "HDFC Bank",
            bankCode = "", // matches any sender, to prove the rule still wins
            transactionType = "other",
            direction = "credit", // deliberately wrong, to prove it's unused
            paymentMethod = null,
            txnCategoryHint = null,
            skeleton = "totally different shape {amount}",
            templateConfidence = 0.9,
        )

        val result = parser.parse(body, templates = listOf(decoy))

        assertEquals("upi_debit", result.matchedRuleId)
        assertNull(result.matchedTemplateId)
        assertEquals("debit", result.direction)
    }

    @Test
    fun `disambiguates between several UPI modes on different accounts by account plus type`() {
        val body = "Sent Rs.500.00\n" +
            "From HDFC Bank A/C *4321\n" +
            "To Test Merchant\n" +
            "On 18-08-26\n" +
            "Ref 999888777666\n" +
            "Not You?\n" +
            "Call 18002586161/SMS BLOCK UPI to 7308080808"

        val accounts = listOf(
            mapOf("id" to "acc-1", "lastSixDigits" to "004321"),
            mapOf("id" to "acc-2", "lastSixDigits" to "009988"),
        )
        val paymentModes = listOf(
            mapOf("id" to "upi-on-acc2", "type" to "upi", "accountId" to "acc-2"),
            mapOf("id" to "upi-on-acc1", "type" to "upi", "accountId" to "acc-1"),
            mapOf("id" to "card-on-acc1", "type" to "debitCard", "accountId" to "acc-1"),
        )

        val result = parser.parse(body, accounts = accounts, paymentModes = paymentModes)

        assertEquals("acc-1", result.matchedAccountId)
        assertEquals("upi-on-acc1", result.matchedPaymentModeId)
    }

    @Test
    fun `an empty string yields zero confidence and no rule match`() {
        val result = parser.parse("")

        assertNull(result.amount)
        assertEquals("none", result.matchedRuleId)
        assertEquals(0.0, result.confidence, 0.0)
        assertTrue(result.isLowConfidence)
    }
}

private fun JSONObject.optNullableDouble(key: String): Double? =
    if (!has(key) || isNull(key)) null else getDouble(key)

private fun JSONObject.optNullableString(key: String): String? =
    if (!has(key) || isNull(key)) null else getString(key)

File: test/services/sms/local_sms_parser_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:ledger/models/account.dart';
import 'package:ledger/models/payment_mode.dart';
import 'package:ledger/models/sms_template.dart';
import 'package:ledger/services/sms/local_sms_parser.dart';

LocalSmsParser _loadParser() {
  final rulesJson = jsonDecode(
    File('assets/sms_patterns/bank_patterns.json').readAsStringSync(),
  ) as Map<String, dynamic>;
  return LocalSmsParser.fromJson(rulesJson);
}

List<Map<String, dynamic>> _loadSamples() {
  final fixture = jsonDecode(
    File('test/fixtures/sms_samples.json').readAsStringSync(),
  ) as Map<String, dynamic>;
  return (fixture['samples'] as List).cast<Map<String, dynamic>>();
}

void main() {
  final parser = _loadParser();
  final samples = _loadSamples();

  group('LocalSmsParser against shared fixtures', () {
    for (final sample in samples) {
      final id = sample['id'] as String;
      final body = sample['body'] as String;
      final expected = sample['expected'] as Map<String, dynamic>;

      test(id, () {
        final result = parser.parse(body);

        expect(result.amount, expected['amount']);
        expect(result.direction, expected['direction']);
        expect(result.paymentMethod, expected['paymentMethod']);
        expect(result.txnCategoryHint, expected['txnCategoryHint']);
        expect(result.referenceNumber, expected['referenceNumber']);
        expect(result.merchantCandidate, expected['merchantCandidate']);
        expect(result.accountLastDigits, expected['accountLastDigits']);
        expect(result.matchedRuleId, expected['matchedRuleId']);

        switch (expected['tier'] as String) {
          case 'high':
            expect(result.isHighConfidence, isTrue,
                reason: 'expected high confidence, got ${result.confidence}');
            break;
          case 'medium':
            expect(result.isMediumConfidence, isTrue,
                reason:
                    'expected medium confidence, got ${result.confidence}');
            break;
          case 'low':
            expect(result.isLowConfidence, isTrue,
                reason: 'expected low confidence, got ${result.confidence}');
            break;
        }
      });
    }
  });

  group('LocalSmsParser learned-template tier', () {
    // A shape no static rule in bank_patterns.json covers at all — it
    // deliberately avoids every rule-triggering keyword ("debited",
    // "credited", "upi", "credit card", "atm", "refund"), same as the
    // real HDFC "Sent"/"Received" format that motivated this tier.
    // Exactly the case the template tier exists for: a bank format
    // previously seen only via the AI fallback (see
    // ClaudeService.parseSmsTransaction), now resolvable locally.
    final icici = SmsTemplate(
      id: 'tpl-icici-debit',
      bank: 'ICICI Bank',
      bankCode: 'ICICIB',
      transactionType: 'bank_debit',
      direction: 'debit',
      paymentMethod: 'bankTransfer',
      txnCategoryHint: null,
      skeleton: 'ICICI Alert: INR {amount} moved from A/C XX{acct_last4} '
          'to {merchant} on {date}. Ref {refno}.',
      templateConfidence: 0.92,
      createdAt: DateTime.utc(2026),
      lastMatchedAt: DateTime.utc(2026),
    );
    const icc = 'ICICI Alert: INR 1250.50 moved from A/C XX7788 to '
        'BLINKIT on 19-Aug-26. Ref 445566778899.';

    test('falls back to a matching template when no static rule fits, '
        'reporting matchedRuleId "template" and the template\'s id', () {
      final result = parser.parse(icc,
          sender: 'AD-ICICIB-S', templates: [icici]);

      expect(result.matchedRuleId, 'template');
      expect(result.matchedTemplateId, 'tpl-icici-debit');
      expect(result.amount, 1250.5);
      expect(result.direction, 'debit');
      expect(result.paymentMethod, 'bankTransfer');
      expect(result.referenceNumber, '445566778899');
      expect(result.merchantCandidate, 'BLINKIT');
      expect(result.accountLastDigits, '7788');
      expect(result.isHighConfidence, isTrue,
          reason: 'amount+direction+ref+paymentMethod+merchant = 80/100');
    });

    test('a template from a different bank\'s sender is not tried', () {
      final result = parser.parse(icc,
          sender: 'VM-HDFCBK-T', templates: [icici]);

      expect(result.matchedRuleId, 'none');
      expect(result.matchedTemplateId, isNull);
    });

    test('a static rule match takes priority over a template, even one '
        'that would also match this exact body', () {
      // upi_debit already covers this shape via bank_patterns.json's own
      // rules — the template tier must never be consulted once a static
      // rule fires, so a bad/duplicate learned template can't override
      // hand-verified extraction.
      const body =
          'Rs.286.00 debited from A/C XX1234 to VPA rajesh@okhdfc on '
          '15-08-26. UPI Ref No 402312345678. Not you? Call 1800123456';
      final decoyTemplate = SmsTemplate(
        id: 'decoy',
        bank: 'HDFC Bank',
        bankCode: '', // matches any sender, to prove the rule still wins
        transactionType: 'other',
        direction: 'credit', // deliberately wrong, to prove it's unused
        skeleton: 'totally different shape {amount}',
        templateConfidence: 0.9,
        createdAt: DateTime.utc(2026),
        lastMatchedAt: DateTime.utc(2026),
      );

      final result = parser.parse(body, templates: [decoyTemplate]);

      expect(result.matchedRuleId, 'upi_debit');
      expect(result.matchedTemplateId, isNull);
      expect(result.direction, 'debit');
    });
  });

  group('LocalSmsParser confidence rubric', () {
    test('matching a known account last-digits adds the instrument bonus', () {
      const body =
          'INR 12000.00 debited from your account for NEFT transfer to A/C XX4321. Avl Bal INR 45000.00';
      final withoutAccount = parser.parse(body);
      final withAccount = parser.parse(
        body,
        accounts: [
          Account(
            id: 'acc-1',
            userId: 'user-1',
            title: 'Checking',
            bankName: 'Test Bank',
            lastSixDigits: '564321',
            balance: 0,
            holderName: 'Test User',
            createdAt: DateTime.utc(2026),
          ),
        ],
      );

      expect(withAccount.matchedAccountId, 'acc-1');
      expect(withAccount.confidence - withoutAccount.confidence, closeTo(0.20, 0.001));
    });

    test('matching a known payment mode last-digits adds the instrument bonus',
        () {
      const body =
          'Rs.4500.00 spent using your Credit Card XX9988 at a merchant on 12-08-26.';
      final withoutMode = parser.parse(body);
      final withMode = parser.parse(
        body,
        paymentModes: [
          PaymentMode(
            id: 'mode-1',
            userId: 'user-1',
            type: PaymentModeType.creditCard,
            title: 'HDFC Card',
            lastFourDigits: '9988',
            createdAt: DateTime.utc(2026),
          ),
        ],
      );

      expect(withMode.matchedPaymentModeId, 'mode-1');
      expect(withMode.confidence - withoutMode.confidence, closeTo(0.20, 0.001));
    });

    test(
        'disambiguates between several UPI modes on different accounts by '
        'account + type, since a UPI id has no digits of its own to match',
        () {
      const body = 'Sent Rs.500.00\n'
          'From HDFC Bank A/C *4321\n'
          'To Test Merchant\n'
          'On 18-08-26\n'
          'Ref 999888777666\n'
          'Not You?\n'
          'Call 18002586161/SMS BLOCK UPI to 7308080808';

      final accounts = [
        Account(
          id: 'acc-1',
          userId: 'user-1',
          title: 'HDFC Savings',
          bankName: 'HDFC Bank',
          lastSixDigits: '004321',
          balance: 0,
          holderName: 'Test User',
          createdAt: DateTime.utc(2026),
        ),
        Account(
          id: 'acc-2',
          userId: 'user-1',
          title: 'ICICI Savings',
          bankName: 'ICICI Bank',
          lastSixDigits: '009988',
          balance: 0,
          holderName: 'Test User',
          createdAt: DateTime.utc(2026),
        ),
      ];
      final paymentModes = [
        PaymentMode(
          id: 'upi-on-acc2',
          userId: 'user-1',
          type: PaymentModeType.upi,
          accountId: 'acc-2',
          title: 'UPI on ICICI',
          createdAt: DateTime.utc(2026),
        ),
        PaymentMode(
          id: 'upi-on-acc1',
          userId: 'user-1',
          type: PaymentModeType.upi,
          accountId: 'acc-1',
          title: 'UPI on HDFC',
          createdAt: DateTime.utc(2026),
        ),
        PaymentMode(
          id: 'card-on-acc1',
          userId: 'user-1',
          type: PaymentModeType.debitCard,
          accountId: 'acc-1',
          title: 'HDFC Debit Card',
          createdAt: DateTime.utc(2026),
        ),
      ];

      final result =
          parser.parse(body, accounts: accounts, paymentModes: paymentModes);

      expect(result.matchedAccountId, 'acc-1');
      expect(result.matchedPaymentModeId, 'upi-on-acc1');
    });

    test('an empty string yields zero confidence and no rule match', () {
      final result = parser.parse('');

      expect(result.amount, isNull);
      expect(result.matchedRuleId, 'none');
      expect(result.confidence, 0.0);
      expect(result.isLowConfidence, isTrue);
    });

    test('tier boundaries are inclusive/exclusive as documented', () {
      expect(
        const LocalParseResult(matchedRuleId: 'none', confidence: 0.80)
            .isHighConfidence,
        isTrue,
      );
      expect(
        const LocalParseResult(matchedRuleId: 'none', confidence: 0.79)
            .isHighConfidence,
        isFalse,
      );
      expect(
        const LocalParseResult(matchedRuleId: 'none', confidence: 0.40)
            .isMediumConfidence,
        isTrue,
      );
      expect(
        const LocalParseResult(matchedRuleId: 'none', confidence: 0.39)
            .isMediumConfidence,
        isFalse,
      );
      expect(
        const LocalParseResult(matchedRuleId: 'none', confidence: 0.39)
            .isLowConfidence,
        isTrue,
      );
    });
  });
}
